Firmware/cmdqueue.cpp

```diff
@@ -478,35 +475,48 @@ void get_command()
             serial_count = 0;
             return;
         }
+        // Handle KILL early, even when Stopped
+        if(strcmp(cmdbuffer+bufindw+CMDHDRSIZE, "M112") == 0)
+          kill(MSG_M112_KILL, 2);
+        // Handle the USB timer
         if ((strchr_pointer = strchr(cmdbuffer+bufindw+CMDHDRSIZE, 'G')) != NULL) {
             if (!IS_SD_PRINTING) {
                 usb_timer.start();
             }
-            if (Stopped == true) {
-                if (code_value_uint8() <= 3) {
-                    SERIAL_ERRORLNRPGM(MSG_ERR_STOPPED);
-                    LCD_MESSAGERPGM(_T(MSG_STOPPED));
-                }
-            }
-        } // end of 'G' command
+        }
+        if (Stopped == true) {
+            // Stopped can be set either during error states (thermal error: cannot continue), or
+            // when a printer-initiated action is processed. In such case the printer will send to
+            // the host an action, but cannot know if the action has been processed while new
+            // commands are being sent. In this situation we just drop the command while issuing
+            // periodic "busy" messages in the main loop. Since we're not incrementing the received
+            // line number, a request for resend will happen (if necessary), ensuring we don't skip
+            // commands whenever Stopped is cleared and processing resumes.
+            serial_count = 0;
+            return;
+        }
+
+        // Command is complete: store the current line into buffer, move to the next line.
 
-        //If command was e-stop process now
-        if(strcmp(cmdbuffer+bufindw+CMDHDRSIZE, "M112") == 0)
-          kill(MSG_M112_KILL, 2);
-        
-        // Store the current line into buffer, move to the next line.
 		// Store type of entry
         cmdbuffer[bufindw] = gcode_N ? CMDBUFFER_CURRENT_TYPE_USB_WITH_LINENR : CMDBUFFER_CURRENT_TYPE_USB;
+
 #ifdef CMDBUFFER_DEBUG
         SERIAL_ECHO_START;
         SERIAL_ECHOPGM("Storing a command line to buffer: ");
         SERIAL_ECHO(cmdbuffer+bufindw+CMDHDRSIZE);
         SERIAL_ECHOLNPGM("");
 #endif /* CMDBUFFER_DEBUG */
+
+        // Store command itself
         bufindw += strlen(cmdbuffer+bufindw+CMDHDRSIZE) + (1 + CMDHDRSIZE);
         if (bufindw == sizeof(cmdbuffer))
             bufindw = 0;
         ++ buflen;
+
+        // Update the processed gcode line
+        gcode_LastN = gcode_N;
```

Comment:
@wavexx Why was this moved here? After this change a command with no N parameter will reset gcode_LastN to 0, breaking serial communication

Reply:
Didn't think of this case.

It was moved later since we need to store only the last valid command received while !Stopped. When removing Stopped we ask a resend from that point.

What is supposed to happen at the protocol level if the stream _was_ using line numbers, but one without is received?

Reply:
In such a case it shouldn't change gcode_LastN. So this would be valid:
```
N5 aaaaaaaaaaa*123
M105
N6 bbbbbbbbb*123
```
With your change this breaks

I got this snippet from Tomas:
```
N1 $1 M201 X10000 Y10000 *53
ok
N2 $2 G28 W *112
tmc2130_home_enter(axes_mask=0x01)
  0 step=30 mscnt= 483
tmc2130_goto_step 0 45 2 1000
tmc2130_home_exit tmc2130_sg_homing_axes_mask=0x01
tmc2130_home_enter(axes_mask=0x02)
echo:busy: processing
  0 step= 2 mscnt=  33
tmc2130_goto_step 1 1 2 1000
tmc2130_home_exit tmc2130_sg_homing_axes_mask=0x02
echo:busy: processing
ok
N3 $3 G91 *5
ok
M105
ok T:32.0 /0.0 B:25.7 /0.0 T0:32.0 /0.0 @:0 B@:0 P:26.2 A:30.1
N4 $4 G1 F3000 Z10 *34
Error:Line Number is not Last Line Number+1, Last Line: 0
Resend: 1
ok
```

Reply:
So it should only update if gcode_N is non-zero? This means we can get inbound commands that will be simply discarded.

For example:

```
N1 G1 ...
G1 ... ; power lost here
```

recovery is broken?

Reply:
Lines with no line number are from USB anyway, so no power panic for now. Also, I think he uses lines without line number only for gcodes triggered from a source different the gcode file. So every "print" gcode still has line numbers, but stuff like M105 and PRUSA FAN don't (or at least I hope they don't)

Reply:
Mmh, I was thinking for the case when a print it sent from serial. The fact that one can downgrade to skip line numbers and checksums is potentially problematic if you have line noise.

But I don't see a problem with this. Can you make a PR for this?

Reply:
Are these the correct changes? I'm not sure if just reverting this breaks anything. I didn't have time to test.
https://github.com/leptun/Prusa-Firmware/tree/Fix_line_numbering

Reply:
Can't attach a comment without a PR.

IMHO the correct fix is to simply to wrap the assignment inside an if(gcode_N > 0). The last gcode number should be ignored if "Stopped" is set, and it's checked a few lines down.

Reply:
Does the protocol support line N0 or is that something we don’t do? If we don’t then that will work and I will change as you say.

Reply:
I don't know this at the moment. But we can change the default value of the variable above to signal that no line number has been seen on the current line instead of assuming that 0 is "no line".

Reply:
Can we use gcode_LastN instead of 0 to initialize gcode_N?